hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+	 */
+	private String fileNameShort;
+
+	/**
+	 * the path to which we write record stream files and signature files
+	 */
+	private final String dirPath;
+	private RecordStreamFile.Builder recordStreamFileBuilder;
+	private int recordFileVersion;
+
+	public RecordStreamFileWriter(
+			final String dirPath,
+			final long logPeriodMs,
+			final Signer signer,
+			final boolean startWriteAtCompleteWindow,
+			final RecordStreamType streamType
+	) throws NoSuchAlgorithmException {
+		this.dirPath = dirPath;
+		this.logPeriodMs = logPeriodMs;
+		this.signer = signer;
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		this.streamType = streamType;
+		this.streamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
+		this.metadataStreamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
+	}
+
+	@Override
+	public void addObject(T object) {
+		if (checkIfShouldWriteNewFile(object)) {
+			// if we are currently writing a file,
+			// finish current file and generate signature file
+			closeCurrentAndSign();
+			// write the beginning of the new file
+			beginNew(object);
+		}
+
+		// if recordStreamFile is null, it means startWriteAtCompleteWindow is true,
+		// and we are still in the first incomplete window, so we don't serialize this object;
+		// so we only serialize the object when stream is not null
+		if (recordStreamFileBuilder != null) {
+			consume(object);
+		}
+
+		// update runningHash
+		this.runningHash = object.getRunningHash();
+	}
+
+	/**
+	 * check whether need to start a new file
+	 *
+	 * @param object
+	 * 		the object to be written into file
+	 * @return whether we should start a new File for writing this object
+	 */
+	public boolean checkIfShouldWriteNewFile(T object) {
+		Instant currentConsensusTimestamp = object.getTimestamp();
+		boolean result;
+		if (lastConsensusTimestamp == null && !startWriteAtCompleteWindow) {
+			// this is the first object, we should start writing it to new File
+			result = true;
+		} else if (lastConsensusTimestamp == null) {
+			// this is the first object, we should wait for the first complete window
+			result = false;
+		} else {
+			// if lastConsensusTimestamp and currentConsensusTimestamp are in different periods,
+			// we should start a new file
+			result =
+					getPeriod(lastConsensusTimestamp, logPeriodMs) != getPeriod(currentConsensusTimestamp, logPeriodMs);
+		}
+		// update lastConsensusTimestamp
+		lastConsensusTimestamp = currentConsensusTimestamp;
+		return result;
+	}
+
+	/**
+	 * if recordStreamFile is not null:
+	 * write last runningHash to current file;
+	 * close current file;
+	 * and generate a corresponding signature file
+	 */
+	public void closeCurrentAndSign() {
+		if (recordStreamFileBuilder != null) {
+			// generate file name
+			final var firstTxnTimestamp =
+					recordStreamFileBuilder.getRecordStreamItems(0).getRecord().getConsensusTimestamp();
+			this.file = new File(
```

Comment:
Move to method variable since it's not used outside this.

Reply:
[Fixed.](https://github.com/hashgraph/hedera-services/pull/3461/commits/5d8e831b26143ae625c5109de968aafa6cf5f022)